Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,337 @@
+  An inductive definition representing that a partial function `f` is recursive in oracle `g`.
+- `turing_reducible`: A relation defining Turing reducibility between partial functions.
+- `turing_equivalent`: A relation defining Turing equivalence between partial functions.
+- `TuringDegree`:
+  The type of Turing degrees, defined as equivalence classes under `turing_equivalent`.
+- `join`: Combines two partial functions into one by mapping even and odd numbers to respective functions.
+
+## Notation
+
+- `f ≤ᵀ g` : `f` is Turing reducible to `g`.
+- `f ≡ᵀ g` : `f` is Turing equivalent to `g`.
+- `f ⊔ g` : The join of partial functions `f` and `g`.
+
+## Implementation Notes
+
+The type of partial functions recursive in an oracle `g` is the smallest type containing
+the constant zero, the successor, projections, the oracle `g`, and is closed under
+pairing, composition, primitive recursion, and μ-recursion. The `join` operation
+combines two partial functions into a single partial function by mapping even and odd
+numbers to the respective functions.
+
+## References
+
+* [Carneiro2018] Carneiro, Mario.
+  *Formalizing Computability Theory via Partial Recursive Functions*.
+  arXiv preprint arXiv:1810.08380, 2018.
+* [Odifreddi1989] Odifreddi, Piergiorgio.
+  *Classical Recursion Theory: The Theory of Functions and Sets of Natural Numbers,
+  Vol. I*. Springer-Verlag, 1989.
+* [Soare1987] Soare, Robert I. *Recursively Enumerable Sets and Degrees*. Springer-Verlag, 1987.
+* [Gu2015] Gu, Yi-Zhi. *Turing Degrees*. Institute for Advanced Study, 2015.
+
+## Tags
+
+Computability, Oracle, Turing Degrees, Reducibility, Equivalence Relation
+-/
+
+open Primrec Nat.Partrec
+
+/--
+The type of partial functions recursive in an oracle `g` is the smallest type containing
+the constant zero, the successor, left and right projections, the oracle `g`, and is closed under
+pairing, composition, primitive recursion, and μ-recursion.
+-/
+inductive RecursiveIn (g : ℕ →. ℕ) : (ℕ →. ℕ) → Prop
+  | zero : RecursiveIn g (fun _ => 0)
+  | succ : RecursiveIn g Nat.succ
+  | left : RecursiveIn g (fun n => (Nat.unpair n).1)
+  | right : RecursiveIn g (fun n => (Nat.unpair n).2)
+  | oracle : RecursiveIn g g
+  | pair {f h : ℕ →. ℕ} (hf : RecursiveIn g f) (hh : RecursiveIn g h) :
+      RecursiveIn g (fun n => (Nat.pair <$> f n <*> h n))
+  | comp {f h : ℕ →. ℕ} (hf : RecursiveIn g f) (hh : RecursiveIn g h) :
+      RecursiveIn g (fun n => h n >>= f)
+  | prec {f h : ℕ →. ℕ} (hf : RecursiveIn g f) (hh : RecursiveIn g h) :
+      RecursiveIn g (fun p =>
+        let (a, n) := Nat.unpair p
+        n.rec (f a) (fun y ih => do
+          let i ← ih
+          h (Nat.pair a (Nat.pair y i))))
+  | rfind {f : ℕ →. ℕ} (hf : RecursiveIn g f) :
+      RecursiveIn g (fun a =>
+        Nat.rfind (fun n => (fun m => m = 0) <$> f (Nat.pair a n)))
+
+/--
+`f` is Turing reducible to `g` if `f` is recursive in `g`.
+-/
+def turing_reducible (f g : ℕ →. ℕ) : Prop :=
+  RecursiveIn g f
```

Comment:
Is this definition needed? Can't you directly use `RecursiveIn`?

Reply:
Having this as shorthand does make things convenient. Do you think it's necessary to remove?

Reply:
I think in the long run either it won't be convenient because either stuff will have to be duplicated with `turing_reducible` and `RecursiveIn`, or everything will use one of these, and then you might as well only define it

Reply:
Fixed in the latest push, thank you